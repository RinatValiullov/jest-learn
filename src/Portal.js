import React, { Component } from "react";
import ReactDOM from "react-dom";

class Portal extends Component {
  el = document.createElement("div").setAttribute("id", "portal");

  componentDidMount() {
    document.body.appendChild(this.el);
  }

  componentWillUnmount() {
    document.body.removeChild(this.el);
  }

  render() {
    return ReactDOM.createPortal(this.props.children, this.el);
  }
}

export default Portal;
