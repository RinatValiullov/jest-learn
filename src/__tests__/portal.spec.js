import { shallow } from "enzyme";
import React from "react";
import ReactDOM from "react-dom";
import Portal from "../Portal";

const componentDidMountSpy = jest.spyOn(Portal.prototype, "componentDidMount");
const componentWillUnmountSpy = jest.spyOn(
  Portal.prototype,
  "componentWillUnmount"
);

const setupComponent = () =>
  shallow(
    <Portal>
      <div>4ront</div>
    </Portal>
  );

describe("Portal component", () => {
  let component;

  beforeEach(() => {
    jest.spyOn(document.body, "appendChild").mockImplementation(() => {});
    jest.spyOn(document.body, "removeChild").mockImplementation(() => {});
    component = setupComponent();
  });

  afterEach(() => {
    document.body.appendChild.mockRestore();
    document.body.removeChild.mockRestore();
  });

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  it("should render Portal component", () => {
    expect(component).toMatchSnapshot();
  });

  it("should call 'appendChild' when component mounted", () => {
    expect(document.body.appendChild).toHaveBeenCalledTimes(1);
  });

  it("should call 'removeChild' when component unmounted", () => {
    component.unmount();
    expect(document.body.removeChild).toHaveBeenCalledTimes(1);
  });
});
