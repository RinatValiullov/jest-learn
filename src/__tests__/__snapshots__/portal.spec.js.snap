// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal component should render Portal component 1`] = `
<div>
  4ront
</div>
`;
